MekHQ/src/mekhq/campaign/personnel/generator/RandomPortraitGenerator.java

```diff
@@ -0,0 +1,141 @@
+        File searchFile = new File(genderFile, Person.getRoleDesc(p.getPrimaryRole(), p.isClanner()));
+
+        possiblePortraits = getPossibleRandomPortraits(existingPortraits, searchFile);
+
+        if (possiblePortraits.isEmpty()) {
+            String searchCat_RoleGroup = "";
+            if (p.isAdminPrimary()) {
+                searchCat_RoleGroup = "Admin";
+            } else if (p.getPrimaryRole() == Person.T_SPACE_CREW
+                    || p.getPrimaryRole() == Person.T_SPACE_GUNNER
+                    || p.getPrimaryRole() == Person.T_SPACE_PILOT
+                    || p.getPrimaryRole() == Person.T_NAVIGATOR) {
+                searchCat_RoleGroup = "Vessel Crew";
+            } else if (p.isTechPrimary()) {
+                searchCat_RoleGroup = "Tech";
+            } else if (p.getPrimaryRole() == Person.T_MEDIC
+                    || p.getPrimaryRole() == Person.T_DOCTOR) {
+                searchCat_RoleGroup = "Medical";
+            }
+
+            // TODO : Java 11 : Swap to isBlank
+            if (!searchCat_RoleGroup.trim().isEmpty()) {
+                searchFile = new File(genderFile, searchCat_RoleGroup);
+                possiblePortraits = getPossibleRandomPortraits(existingPortraits, searchFile);
+            }
+        }
+
+        if (possiblePortraits.isEmpty()) {
+            searchFile = new File(genderFile, p.hasPrimaryCombatRole() ? "Combat" : "Support");
+            possiblePortraits = getPossibleRandomPortraits(existingPortraits, searchFile);
+        }
+
+        if (possiblePortraits.isEmpty()) {
+            possiblePortraits = getPossibleRandomPortraits(existingPortraits, genderFile);
+        }
+
+        if (!possiblePortraits.isEmpty()) {
+            String chosenPortrait = possiblePortraits.get(Compute.randomInt(possiblePortraits.size()));
+            String[] temp = chosenPortrait.split(":");
+            if (temp.length == 2) {
+                return new Portrait(temp[0], temp[1]);
+            } else {
+                MekHQ.getLogger().error("Failed to generate portrait for " + p.getFullTitle() + ". "
+                        + chosenPortrait + " does not split into an array of length 2.");
+            }
+        } else {
+            MekHQ.getLogger().warning("Failed to generate portrait for " + p.getFullTitle()
+                    + ". No possible portraits found.");
+        }
+
+        return new Portrait();
+    }
+
+    /**
+     * This is a helper method that determines what possible unassigned portraits can be generated
+     * based on the supplied subdirectory
+     *
+     * @param existingPortraits the list of existing portraits that have already been assigned
+     * @param subDir the subdirectory to search
+     * @return a list of all possible unassigned random portraits
+     */
+    private static List<String> getPossibleRandomPortraits(List<String> existingPortraits, File subDir) {
```

Comment:
Existing Portraits should be a Set to avoid bad runtimes on large campaigns.

Reply:
Pre-existing and I've never heard any issues with its performance before even on the largest campaigns. However, given it is for contains it does make sense to swap over, so I've done so.